src/librustc/traits/error_reporting.rs

```diff
@@ -103,16 +105,191 @@ pub fn report_projection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
     }
 }
 
+fn impl_substs<'a, 'tcx>(fcx: &InferCtxt<'a, 'tcx>,
+                         did: DefId,
+                         obligation: PredicateObligation<'tcx>)
+                         -> subst::Substs<'tcx> {
+    let tcx = fcx.tcx;
+
+    let ity = tcx.lookup_item_type(did);
+    let (tps, rps, _) =
+        (ity.generics.types.get_slice(subst::TypeSpace),
+         ity.generics.regions.get_slice(subst::TypeSpace),
+         ity.ty);
+
+    let rps = fcx.region_vars_for_defs(obligation.cause.span, rps);
+    let mut substs = subst::Substs::new(
+        subst::VecPerParamSpace::empty(),
+        subst::VecPerParamSpace::new(rps, Vec::new(), Vec::new()));
+    fcx.type_vars_for_defs(obligation.cause.span, subst::ParamSpace::TypeSpace, &mut substs, tps);
+    substs
+}
+
+trait AssociatedWeight {
+    fn get_weight(&self) -> usize;
+}
+
+impl<'a> AssociatedWeight for TypeVariants<'a> {
+    // First number is for "global" weight and second number is for bigger precision
+    fn get_weight(&self) -> usize {
+        match *self {
+            TypeVariants::TyBool => 11,
```

Comment:
okay, if i understand the goal is correctly, these numbers are partitioning the kinds of type-variants into first a coarse set of equivalence classes: {Bool, Char, Str} + {Numerics, RawPtr} + Records + Sequences + et cetera, and then the one's place is distinguishing between the elements of each equivalence class.

This way, a coarse grained weight computation will lump `bool`+`char`+`str` together, since they are are all quite different from a number-like thing like `i32` or `f64`, but if that coarse computation doesn't find a unique candidate `impl`, then it resorts to the finer grain filter on the candidates that share the same coarse weight.

The numbers seem like they are meant to carry some meaning beyond the equivalence class that they establish -- i.e. if I understand what these numbers are saying: `bool` is somewhat like `i32`, while `i32` is somewhat like an `enum`, but a `bool` is quite different from an `enum`?  And all of them are very very different from a closure?

(Please correct me if any of the above is incorrect.)

Some comments:

 * In terms of code presentation, I would add new lines between each set of equivalence classes. (E.g. add a line break between `TyStr` and `TyInt`, to mark the divide between a coarse weight of `1` and `2`.)

 * I am not quite clear on some of the choices made here. For example, why give `TyRef` the same coarse weight (5) that  `TyFnDef`+`TyFnPtr` get?

 * I can imagine many small differences accumulating to eventually become larger than one big difference in weight. That seems like it is by design (i.e. in the end if many types do not match then it seems between to suggest the impl that has only one difference). But does it mean that you should use larger numbers for some of the entries at the end of the `match` below? As an example, does it ever make sense to suggest an `impl` with `TyInfer` or `TyError`, simply because there happened to be >10 small differences in the other candidates?

----

Overall it seems a bit ad hoc, but it also seems like it gets